### ISSUE TITLE
Summarizer is Broken

### ISSUE THREAD

**Issue by erniejunior** [reinforceio/tensorforce#415]
I am opening a _summary_ issue (sorry could not resist) here for the fact, that the summarizer is broken. This is a shame because the summaries could be very helpful for debugging.

See also the issues #396 and #350 created earlier this year. @AlexKuhnle mentioned, that `tf.contrib.summary` [is the way to go](https://github.com/reinforceio/tensorforce/issues/396#issuecomment-381320982) but then [two days later was not sure any more](https://github.com/reinforceio/tensorforce/issues/396#issuecomment-381742747) after seemingly failing to quickly do the port.

So I tried the same and also failed :-) but maybe somebody else wants to continue my half-done work? Feel free to grab my [feature branch](https://github.com/erniejunior/tensorforce/tree/feature/summary_v2) if it helps. Unfortunately I do not have the time to continue on that right now.

Some pointers if you decide to go this route:
* Code (including documentation which is not readable in the official tensorflow web page rendering) [for tf.contrib.summary](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/summary/summary.py)
* Code (and documentation) for [session hooks](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/training/basic_session_run_hooks.py).
* When you look into my feature branch you can see that I mostly just renamed tf.summary to tf.contrib.summary where it applies and I am using the `UpdateSummarySaverHook` in `util.py` to set up the summary system. It does not crash (but also does not produce any output) for the `inputs` and `gradients` labels and crashes as soon as the `losses` label is enabled.

**Comment by nskplta** [comment]
Hi, any update on this? Many of us are expecting a solution, thanks for the help!

**Comment by erniejunior** [comment]
Nope :-( still did not get around to look deeper into it. Just view it as a black box and optimize the hyperparameters using SMAC or hyperopt ;-)

**Comment by erniejunior** [comment]
This was not meant to be taken too serious. Basically what I meant is: instead of looking into your system internals (using the summaries) and look what is wrong with it to decide what parameters and settings to change, just randomly play around with them until it works. This random playing around with settings is called hyperparameter optimization and there exist toolkits do do that. [SMAC](http://www.ml4aad.org/algorithm-configuration/smac/) and [hyperopt](http://hyperopt.github.io/hyperopt/) are two popular ones.

**Comment by madpower2000** [comment]
Please, please fix summarizer! Unfortunately without it tensorforce mostly useless piece :( By the way, as I understood,  summarizer was starting crashing after new version tensorflow was rollout, but  what exactly last know tensorflow version still able to run summarizer?

**Comment by AlexKuhnle** [comment]
Hehe, @madpower2000, that's quite a harsh judgement. ;-)   The problem is that the old TensorFlow summarizers struggle with control flow stuff, which TensorForce makes extensive use of, and while the new summarizer framework is supposed to work better, it still seems to struggle with some bits. I want to assess the new situation soon and will let you know.

**Comment by madpower2000** [comment]
Sorry for harsh judgement, I didn’t mean it. I just was excited, when I discover yours excellent framework, guys! Thanks a lot! Awesome!
I just hoped if I will downgrade TensorFlow version then Summarizer will back to work. 
But any way, is it another way control DNN convergence?

**Comment by AlexKuhnle** [comment]
Was joking with the "harsh judgement" ;-)  it would definitely be a great feature, and it's really annoying that it seems to be TensorFlow which causes problems here (it definitely did with the old summarizer framework). Downgrading doesn't help here, things are supposed to be better now. Downgrading TensorForce would potentially help, but that's not really adviseable either. I sometimes work with `tf.Print`, but that requires modifying internal code, so not really useful. So unfortunately I don't have good advice right now, and just hope that I will get summarizers to work in the next days.

**Comment by gerardsimons** [comment]
Sorry I am not fixing it after all. I think the work done here is awesome, and I wouldn't mind contributing but seeing that I am not getting a whole lot of support (it took 3 weeks before getting any reply on my message listed above) I have decided to switch to [rllib](https://ray.readthedocs.io/en/latest/rllib.html) for now. Perhaps it is of interest to you too @madpower2000

**Comment by AlexKuhnle** [comment]
Sorry for that, @gerardsimons, I wasn't active in support for some months for various reasons. I am in the process of looking into the unanswered issues that built up over the time. If you are interested in getting involved again, let me know.

**Comment by madpower2000** [comment]
This is another treat! Thanks for link @gerardsimons!
But I will be keep eye at reinforce with working summarizers anyway, @AlexKuhnle! ;)

**Comment by AlexKuhnle** [comment]
@all, the summarizers should work again with the latest commit, and the code now uses the new `tf.contrib.summary` framework. It's now relatively easy to add own summarizers in the code, just decide for a tag to be added to `summary_labels` and add the corresponding `tf.contrib.summary` command, as is done, for instance, [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/memory_model.py#L423):
```
if 'losses' in self.summary_labels or 'total-loss' in self.summary_labels:
    tf.contrib.summary.scalar(name='total-loss', tensor=loss)
```
I haven't thoroughly checked existing labels, but preliminary tests worked well. Let me know if you encounter any problems.
